velox/exec/prefixsort/PrefixSortEncoder.h

```diff
@@ -28,51 +28,117 @@ namespace facebook::velox::exec::prefixsort {
 /// Provides encode/decode methods for PrefixSort.
 class PrefixSortEncoder {
  public:
-  /// 1. Only int64_t is supported now.
-  /// 2. Encoding is compatible with sorting ascending with no nulls.
+  /// Encode native primitive types(such as byte, int16, int32, int64, double,
+  /// float) to dest. There will be a separate method to handle string type as
+  /// follow-up.
+  /// 1. The first byte of the encode result is nullByte, which is determined by
+  ///    the nullsFirst :
+  ///    -If nullsFirst : invalid(means null) -> 0, valid(means not null) -> 1
+  ///    -If !nullsFirst : invalid(means null) -> 1, valid(means not null) -> 0
+  /// 2. The remaining bytes are the encoding result of value:
+  ///    -If isNull, we ignore the input value (this means the caller
+  ///     can pass any value, the reason for not using std::optional here is to
+  ///     avoid the overhead of box-unboxing) and set the remaining sizeof(T)
+  ///     bytes to '0', they do not affect the comparison results at all.
+  ///    -If !isNull, we set the remaining bytes as the result of
+  ///     encodeNoNulls, which will consider whether it is ascending or not. The
+  ///     result in descending order is equal to invert-bits(the result in
+  ///     ascending order)
   template <typename T>
-  static FOLLY_ALWAYS_INLINE void encode(T value, char* row);
+  static FOLLY_ALWAYS_INLINE void
+  encode(bool ascending, bool nullsFirst, bool isNull, T value, char* dest);
```

Comment:
I wonder if it makes sense to make the encoder stateful and pass ascending and nullsFirst flags to the constructor.